src/search/merge_and_shrink/labels.h

```diff
@@ -4,45 +4,65 @@
 #include <memory>
 #include <vector>
 
-namespace utils {
-class LogProxy;
-}
-
 namespace merge_and_shrink {
-class Label {
-    /*
-      This class implements labels as used by merge-and-shrink transition systems.
-      Labels are opaque tokens that have an associated cost.
-    */
-    int cost;
+class LabelsConstIterator {
+    const std::vector<int> &label_costs;
+    std::size_t current_index;
+
+    void next_valid_index();
 public:
-    explicit Label(int cost_)
-        : cost(cost_) {
+    LabelsConstIterator(const std::vector<int> &label_costs, bool end);
+    void operator++();
+
+    int operator*() const {
+        return int(current_index);
     }
-    ~Label() {}
-    int get_cost() const {
-        return cost;
+
+    bool operator==(const LabelsConstIterator &rhs) const {
+        return current_index == rhs.current_index;
+    }
+
+    bool operator!=(const LabelsConstIterator &rhs) const {
+        return current_index != rhs.current_index;
     }
 };
 
 /*
-  This class serves both as a container class to handle the set of all labels
-  and to perform label reduction on this set.
+  This class serves both as a container class to handle the set of all
+  labels and to perform label reduction on this set.
+
+  Labels are identified via integers indexing label_costs, which store their
+  costs. When using label reductions, reduced labels are set to -1 in label
+  costs.
```

Comment:
"which store" => "which stores"? I wouldn't use plural to refer to a variable, even if that variable is a collection.  I wouldn't say "reduced labels" to denote label IDs that become unused. In my view, label reduction is like abstraction; if I map labels A and B to the new label C (or to the label A), then in all cases "A" and "B" are reduced. But I think you mean here label IDs that have become obsolete. I would use a word like "labels that become unused".

I would not write "in label costs" to refer to the variable "label_costs". Better to use "label_costs".